hw/drivers/sensors/lis2dw12/src/lis2dw12.c

```diff
@@ -2695,18 +2798,107 @@ lis2dw12_sensor_handle_interrupt(struct sensor *sensor)
          * We need to read this register only if we are
          * interested in the sleep event
          */
-         rc = lis2dw12_get_int_status(itf, &int_status);
-         if (rc) {
-             LIS2DW12_ERR("Could not read int status err=0x%02x\n", rc);
-             return rc;
-         }
+        rc = lis2dw12_get_int_status(itf, &int_status);
+        if (rc) {
+            LIS2DW12_ERR("Could not read int status err=0x%02x\n", rc);
+            return rc;
+        }
+
+        notif_cfg = lis2dw12_find_notif_cfg_by_event(SENSOR_EVENT_TYPE_SLEEP,
+                                                     &lis2dw12->cfg);
+        if (!notif_cfg) {
+            rc = SYS_EINVAL;
+            goto err;
+        }
 
-         if (int_status & LIS2DW12_STATUS_SLEEP_STATE) {
-             /* Sleep state detected */
-             sensor_mgr_put_notify_evt(&lis2dw12->pdd.notify_ctx,
-                                       SENSOR_EVENT_TYPE_SLEEP);
-             STATS_INC(g_lis2dw12stats, sleep_notify);
-         }
+        if (int_status & notif_cfg->notif_src) {
+            /* Sleep state detected */
+            sensor_mgr_put_notify_evt(&lis2dw12->pdd.notify_ctx,
+                                      SENSOR_EVENT_TYPE_SLEEP);
+            lis2dw12_inc_notif_stats(SENSOR_EVENT_TYPE_SLEEP);
+        }
+    }
+
+    rc = lis2dw12_get_sixd_src(itf, &sixd_src);
+    if (rc) {
+        LIS2DW12_ERR("Could not read sixd src err=0x%02x\n", rc);
+        goto err;
+    }
+
+    notif_cfg = lis2dw12_find_notif_cfg_by_event(SENSOR_EVENT_TYPE_ORIENT_CHANGE,
+                                                 &lis2dw12->cfg);
+    if (!notif_cfg) {
+        rc = SYS_EINVAL;
+        goto err;
+    }
+
+    if (sixd_src & notif_cfg->notif_src) {
+
+        /* Orientation change detected, can be a combination of ZH, ZL, YH,
+         * YL, XH and XL
+         */
+
+        sensor_mgr_put_notify_evt(&lis2dw12->pdd.notify_ctx,
+                                  SENSOR_EVENT_TYPE_ORIENT_CHANGE);
+
+        lis2dw12_inc_notif_stats(SENSOR_EVENT_TYPE_ORIENT_CHANGE);
+    }
+
+    notif_cfg = lis2dw12_find_notif_cfg_by_event(SENSOR_EVENT_TYPE_ORIENT_X_CHANGE,
+                                                 &lis2dw12->cfg);
+    if (!notif_cfg) {
+        rc = SYS_EINVAL;
+        goto err;
+    }
+
+    if (sixd_src & notif_cfg->notif_src) {
+
+        /* Orientation change detected, can be a combination of ZH, ZL, YH,
+         * YL, XH and XL
+         */
+
+        sensor_mgr_put_notify_evt(&lis2dw12->pdd.notify_ctx,
+                                  SENSOR_EVENT_TYPE_ORIENT_X_CHANGE);
+
+        lis2dw12_inc_notif_stats(SENSOR_EVENT_TYPE_ORIENT_X_CHANGE);
+    }
+
+    notif_cfg = lis2dw12_find_notif_cfg_by_event(SENSOR_EVENT_TYPE_ORIENT_Y_CHANGE,
```

Comment:
I don't know if this is something worth worrying about or not, but I wonder if there is an efficiency concern here.  Each call to `lis2dw12_find_notif_cfg_by_event` results in a linear traversal of the notification array.  This function (`lis2dw12_find_notif_cfg_by_event`) calls the find function quite a few times (nine by my count).  Like I said, I don't know if this matters in practice.  If it does, it seems it would be more efficient to do a single walk of the notification list, processing each element as they are encountered.  I think it depends on how frequently this function gets called.